Bluetooth: Controller: Schedule BIG event after Periodic Adv event

cvinayak opened the pull request:
Add implementation to find free slot after Periodic
Advertising event, to place the BIG event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor scheduling function depth

Refactor the implementation to get free slot after a
state/role so that the code depth is reduced.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor free slot after role implementation

Refactor the implementation to get free slot after a
state/role so that it is easy for reused for scheduling
Broadcast ISO events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use conn interval unit for Auxiliary event

Use connection interval units of 1250 us for periodic
scheduling of Extended Advertising auxiliary PDU events so
that auxiliary events can be periodically grouped alongwith
the Periodic Advertising Events. This will permit
mitigating overlaps amongst them.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix scheduling of Periodic Advertising events

Fix scheduling of Periodic Advertising events to be after
the group of Auxiliary set events. Do not reduce the slot
offsets as time reservations in ticker for non low latency
variant of controller does not include the slot offsets in
the reservation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Schedule BIG event after Periodic Adv event

Add implementation to find free slot after Periodic
Advertising event, to place the BIG event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Expose the Periodic Adv time reservation func

Make the Periodic Advertising time reservation calculation
function internal public so that advanced scheduling can
use it to scheduling the BIG event after the Periodic
Advertising radio events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1409,16 +1409,26 @@ uint8_t ll_adv_enable(uint8_t enable)
+						  ADV_INT_UNIT_US) +
+						 HAL_TICKER_TICKS_TO_US(
+							ULL_ADV_RANDOM_DELAY),
+						 CONN_INT_UNIT_US);
```
This function is hard to get an overview of and understand (it is a stunning 800 lines of code!), so please enlighten me: Is this change related to connectable ADV, or are you reusing `CONN_INT_UNIT_US` as that is the same unit (1.25 ms)?

cvinayak replied:
Add comment.

Thalley replied:
I got to admit that I do not understand the comment that well. 

I understand that using the unit of 1.25ms groups it with ACL and periodic advertising events, because they share the same unit of interval (1.25ms), but I don't understand how a unit can represent a behavior: 

```
Use connection interval units to represent the periodic behavior of scheduling
```

Would it make sense to create a `PER_ADV_INT_UNIT_US` as well? At least the semantic meaning of the unit will be there.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -769,6 +786,36 @@ void ull_adv_sync_release(struct ll_adv_sync_set *sync)
 	sync_release(sync);
 }
 
+uint32_t ull_adv_sync_time_get(struct ll_adv_sync_set *sync, uint8_t pdu_len)
```
```suggestion
uint32_t ull_adv_sync_time_get(const struct ll_adv_sync_set *sync, uint8_t pdu_len)
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -50,129 +58,78 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+					adv_sync_ull_hdr_get_cb, ticks_anchor,
+					&ticks_to_expire, &ticks_slot);
+	if (ticker_id != 0xff) {
+		struct ll_adv_sync_set *sync;
```
```suggestion
		const struct ll_adv_sync_set *sync;
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -50,129 +58,78 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
 #endif
 #endif /* CONFIG_BT_CTLR_CONN_PARAM_REQ */
+
+static uint8_t after_xxxx_slot_get(uint8_t user_id, uint32_t ticks_slot_abs,
```
Curious name. Can you elaborate on the reasoning behind using `xxxx`?

cvinayak replied:
I could not settle for a name, earlier function was to find a time slot after master, but now the function is refactors to find a time slot after any scheduled ticker. Would `after_match_slot_get`  be a better function name?

Thalley replied:
Possibly. I don't have a good answer for this function name :)

Regarding `master`, I assume that refers to `_mstr_`. Has that not been renamed in the core spec (if relevant to that) by the inclusive naming change? Looks like it has been overlooked.

cvinayak replied:
yes, it was overlooked, updated now.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+
+	ticks_slot_abs += HAL_TICKER_US_TO_TICKS(EVENT_JITTER_US << 3);
+
+	ticker_id = ticker_id_prev = 0xff;
```
Since you are adding the `U` suffix on the others, might as well add that here too: 

```suggestion
	ticker_id = ticker_id_prev = 0xffU;
```

cvinayak replied:
replaced with `TICKER_NULL`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+	ticks_to_expire = ticks_to_expire_prev = 0U;
+	ticks_slot_abs_prev = 0U;
+	while (1) {
+		uint32_t ticks_slot_abs_curr = 0;
```
```suggestion
		uint32_t ticks_slot_abs_curr = 0U;
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+			}
+		}
+
+		success = (ret_cb == TICKER_STATUS_SUCCESS);
+		LL_ASSERT(success);
```
```suggestion
		LL_ASSERT(ret_cb == TICKER_STATUS_SUCCESS);
```

And remove the single use variable

cvinayak replied:
as `ret_cb` is volatile, it cannot be used inside `LL_ASSERT`, it is a coverity rule related to use of volatile variable that it think would change when being checked.

Thalley replied:
Can you please add that as a comment then? That is useful information to know if anyone ever tries to modify the code :)

cvinayak replied:
comment added in code.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+		success = (ret_cb == TICKER_STATUS_SUCCESS);
+		LL_ASSERT(success);
+
+		if (ticker_id == 0xff) {
```
```suggestion
		if (ticker_id == 0xffU) {
```

cvinayak replied:
replaced with `TICKER_NULL`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+
+	ticks_slot_abs += HAL_TICKER_US_TO_TICKS(EVENT_JITTER_US << 3);
+
+	ticker_id = ticker_id_prev = 0xff;
```
What does the value `0xFF` represent?

cvinayak replied:
replaced with `TICKER_NULL`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+#if defined(CONFIG_BT_CTLR_LOW_LAT)
+#if defined(CONFIG_BT_CTLR_XTAL_ADVANCED)
+		if (hdr->ticks_prepare_to_start & XON_BITMASK) {
+			uint32_t ticks_prepare_to_start =
```
```suggestion
			const uint32_t ticks_prepare_to_start =
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+		} else
+#endif /* CONFIG_BT_CTLR_XTAL_ADVANCED */
+		{
+			uint32_t ticks_prepare_to_start =
```
```suggestion
			const uint32_t ticks_prepare_to_start =
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+
+		ticks_slot_abs_curr += hdr->ticks_slot;
+
+		if ((ticker_id_prev != 0xff) &&
```
```suggestion
		if ((ticker_id_prev != 0xffU) &&
```

cvinayak replied:
replaced with `TICKER_NULL`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -596,19 +552,161 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+		ticks_slot_abs_prev = ticks_slot_abs_curr;
+	}
+
+	if (ticker_id_prev != 0xff) {
```
```suggestion
	if (ticker_id_prev != 0xffU) {
```

cvinayak replied:
replaced with `TICKER_NULL`

cvinayak committed:
Bluetooth: Controller: Use define TICKER_NULL instead of 0xFF

Use the define TICKER_NULL to represent invalid ticker id.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Rename use of mstr word to cen for Central

Rename the use of mstr word in advanced scheduling
implementation to cen as a representation for Central
role.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
